Lazy Hook I: A Not Really New Hope

leycec opened the pull request:
Greetings again, earthbound meat suits.

Welcome to this first of three pull requests implementing [codewarrior0](https://github.com/codewarrior0)'s lazy hook loading scheme. I call it "The Lazy Hook Trilogy." You may call it "really, really pretentious."

In this first pull request, we:

* Define a new `PyInstaller.building.imphook.ModuleHook` class encapsulating a lazy loadable hook.
* Define a new `PyInstaller.building.imphook.ModuleHookCache` class caching all lazy loadable hooks.
* Refactor the `PyInstaller.building.build_main.Analysis.assemble()` method to leverage these new classes.

Let's shamble to it, shall we?

## Cracking Open the Shell of History

Six months ago, [tallforasmurf](http://github.com/tallforasmurf) noted my solution of isolating different hook functions to different hook scripts in different hook subdirectories and found it to be: **(A)** balls crazy, **(B)** overly complex, and **(C)** difficult to document, understand, and maintain. (Did we mention "balls crazy"?)

I emphatically agreed. I also had no idea what to do about it. Then [codewarrior0](http://github.com/codewarrior0) proposed a novel remedy: **lazy hook loading.** In [matysek](http://github.com/matysek)'s [proposed improvements](https://github.com/pyinstaller/pyinstaller/pull/1651) to `excludedimports`, codewarrior0 [reproposed](https://github.com/pyinstaller/pyinstaller/pull/1651#issuecomment-153747323) this remedy.

I emphatically agreed... _again._ The time has come to engender this beautiful dream.

## What's So Lazy About Lazy Loading?

Currently, hook scripts are temporarily loaded into memory as `ImportHook` objects directly at their point of use and then discarded. Instead, `ImportHook` objects could be pre-cached for all available hook scripts _without_ actually loading those scripts into memory. On the first use of an `ImportHook` object, that object could then silently load and retain the corresponding hook script into memory for efficient reuse.

Despite being called at different times during application analysis, all hook functions (e.g., `hook()`, `pre_find_module_path()`, `pre_safe_import_module()`) could then be moved into the existing hook scripts in the `PyInstaller/hooks/` directory and all subdirectories of that directory removed entirely.

It is nice.

## New Class #1: `ModuleHook`

The new `ModuleHook` class encapsulates **lazy loading for a single hook.** It defines a variety of private stuff (_the less said, the better for all_) and the following public stuff:

* `code`, a read-only property providing the interpreted contents of this hook's script. Here's where the bloated magic begins. On:
  * First accessing this property, this hook's script is dynamically loaded into an in-memory code object accessible via this property.
  * Subsequently accessing this property, the previously loaded code object is returned.
* `post_graph()`, a method running the "post-graph hook" (i.e., `hook()` function called _after_ `ModuleGraph` construction) for this hook if any.

Laziness shall set you free! _Maybe._

## New Class #2: `ModuleHookCache`

The new `ModuleHookCache` class encapsulates **all lazy loadable hooks for all modules,** mapping module names to lists of `ModuleHook` instances. This class parallels the behaviour of the existing non-lazy-loading `HooksCache` class – except with more awesome.

Here's how this shakes down. For each hook script (either official or custom) discovered by PyInstaller at startup, a `ModuleHook` instance encapsulating this hook is appended to the list of all hooks for the corresponding module in the current `ModuleHookCache`. This cache may then be operated on elsewhere in the codebase.

At the moment, we only iterate this cache to run post-graph hooks after `ModuleGraph` construction. In the frighteningly near future, we'll also perform dictionary lookups on this cache to run both pre-find module path hooks _and_ pre-safe import module hooks during `ModuleGraph` construction.

One small step for leycec; one even smaller step for PyInstaller.

## Newly Refactored: `Analysis.assemble()`

Finally, the `Analysis.assemble()` method has been refactored to leverage these classes in place of their non-lazy-loading variants `HooksCache` and `ImportHook` as follows:

1. The `ModuleHookCache` singleton (appropriately named – _wait for it_ – `module_hook_cache`) is instantiated immediately _before_ `ModuleGraph` construction, mapping all hooked modules to lists of lazy-loadable `ModuleHook` instances. Critically, note that **no hooks have actually been loaded from disk yet.**
1. This singleton is iterated immediately _after_ `ModuleGraph` construction to run post-graph hooks by calling the `post_graph()` method of all lazy-loadable `ModuleHook` instances whose corresponding modules were imported (i.e., added to the `ModuleGraph` during construction). Critically, note that **all such hooks will be implicitly loaded from disk here.**

"Implicitly" is the $1 keyword here.

## How You Too Can Profit from Lazy Hooks in Just 20,143 Easy Steps

Hooks no longer need to be explicitly loaded; they're magically loaded into existence when they need to be. (Isn't that nice? Don't answer that.)

At any time _after_ the instantiation of the `module_hook_cache` singleton in the `Analysis.assemble()` method, the contents of any hook are safely accessible and modifiable via that hook's `code` property. "At any time" means just that: before, during, and after `ModuleGraph` creation and all points in-between.

Let's say you wanted to collect the `excludedimports` attributes defined by the hooks of all importable modules into some giant list for subsequent processing (AKA, a dark and nefarious purpose). Here's how you'd do that in the `Analysis.assemble()` method:

```
# Giant list of all excluded imports.
all_excludedimports = []

# For each hooked module and corresponding list of hooks...
for module_name, module_hooks in module_hook_cache.items():
    # To improve space and time efficiency, ignore unimportable modules!
    # (Optional, but recommended. is_module_importable() is an imaginary
    # function, so insert uncontrolled hand-waving here.)
    if not is_module_importable(module_name):
        continue

    # For each hook for this importable module...
    for module_hook in module_hooks:
        # Record this hook's excluded imports if any. If this hook has not
        # already been lazily loaded, it will be now!
        all_excludedimports.extend(module_hook.code.excludedimports)
```

## The Future (_Now with More Shiny_)

I foresee a bright, new future for PyInstaller! Let's see what that's about.

### Lazy Hook II: Our Ill-fated Quest Continues

The second pull request in this trilogy will **unify post-graph hooks and pre-find module path hooks.** For each pre-find module path hook in the `PyInstaller/hooks/pre_find_module_path/` directory:

* If a post-graph hook with the same basename exists in the `PyInstaller/hooks/` directory:
  1. The contents of this pre-find module path hook will be appended onto the contents of this post-graph hook.
  1. This pre-find module path hook will be removed.
* Else:
  1. This pre-find module path hook will be moved into the `PyInstaller/hooks/` directory as is.

The `PyInstaller/hooks/pre_find_module_path/` directory will then be removed entirely. If PyInstaller is passed a custom hooks directory via the `--additional-hooks-dir` option containing a `pre_find_module_path/` subdirectory, a non-fatal warning informing the user that this subdirectory has been deprecated will be printed. <sup>**[Note:** I highly doubt any user applications _actually_ define custom pre-find module path hooks. However, I also doubted any U.S. president in the past thirty years would be reelected to a second term. Yet here we are.**]**</sup>

Thanks to lazy hook loading, this refactoring _should_ be mostly trivial. But it's the tedious sort of trivial, so I'm deferring it for a bit.

Following this refactoring, pre-find module path hooks will simply be functions named `pre_find_module_path_hook()` in unified hook scripts. A separate directory with separate scripts will be a relic of our once-shameful past, spoken of only in hushed whispers on archived GitHub threads.

### Lazy Hook III: The Wrath of Python Khan

The third and final pull request in this trilogy will **unify post-graph hooks and pre-safe import module hooks** – in the same exact way. It will also remove the existing non-lazy-loading `HooksCache` and `ImportHook` classes, which have been temporarily preserved for posterity until then.

Nothing lasts... but nothing is lost.

leycec commented:
`test_pandas_extension()` appears to be the only failing test. Presumably irrelevant? I, for one, shall ignore the fat panda in the room.

codewarrior0 commented on PyInstaller/building/imphook.py:
```diff
@@ -12,18 +12,449 @@
+                del self[module_name]
+
+
+#FIXME: For space efficiency, convert to use slots rather than "__dict__".
```
Probably not worth the effort unless the hook files number in the thousands or more.

leycec replied:
Infinity is just another number!

But... _Yeezus,_ you're right. Slots smells like premature optimization. I'll gut this `FIXME` comment.

codewarrior0 commented on PyInstaller/building/imphook.py:
```diff
@@ -12,18 +12,449 @@
+    ## Properties
+
+    @property
+    def code(self):
```
Because this isn't actually a code object, and other parts of pyinstaller (and modulegraph) do deal with code objects in various ways, I'd rename it to something like `module` or `namespace` since it is really the namespace of the hook file's module that is being accessed. 

<sub>The hook file does have a code object, but it is executed somewhere inside the import function using the module's namespace as its globals dict.</sub>

leycec replied:
**Urgh.** I foolishly believed this was a bonafide code object. Well, that puts a damper on the day.

`namespace` was my fallback. It's a bit long, though. I really want everyone to start using this everywhere. `ns`, maybe? But acronyms! I shall ponder and revise.

codewarrior0 replied:
A few more names to consider: `names, contents, scope, attributes, attrs, globals, stuff (or schtoff), definitions, defs, variables, vars`

I couldn't find my thesaurus, but the object's literal type is `<type 'module'>`.

htgoebel replied:
Stip of `code` at all. It just adds a useless level to `self.code.excludedimports`. Or did I miss something?
You can simply use `__getattr__` for catch the first access.

leycec replied:
@htgoebel You missed nothing! I've contemplated doing just that. It's magic and convenience, and both of those things are good.

But it _also_ invites namespace collisions (e.g., a hook defining a `post_graph()` or `module_name` attribute). And that is _not_ good. Since we have no control over the contents of custom hooks, I'd prefer segregating hook namespaces to a separate property rather than congeal hook namespaces into an unholy matrimony with the `ModuleHook` namespace.

**But I shall defer to the wisdom of community consensus.** What say you, codewarrior0 and matysek?

leycec replied:
> A few more names to consider: `names`, `contents`, `scope`, `attributes`, `attrs`, `globals`, `stuff` (or `schtoff`), `definitions`, `defs`, `variables`, `vars`

O.K.; those are all uniformly awesome. Thanks for concocting this up, for it has assisted me. `stuff` and `schtoff` are particularly hilarious, albeit mayhap not so unambiguous. `attrs` is attractively precise, but ultimately...

**I dig `scope`.** It's concise; it's unambiguous; it's unabbreviated. What does the rabble think?

codewarrior0 replied:
> What say you, codewarrior0 and matysek?

I agree about segregating the hook's namespace. Not just for the sake of custom hooks, but also so we don't have to worry about whether a hook-global name will conflict with a `ModuleHook` attribute while writing hooks. Ultimately reducing our cognitive load.

> **I dig `scope`.**

Me too. My second choice would be `globals`.

codewarrior0 commented on PyInstaller/building/imphook.py:
```diff
@@ -12,18 +12,449 @@
 logger = logging.getLogger(__name__)
 
 
+class ModuleHookCache(dict):
```
Inheriting from the built-in collection types (`dict`, `list`, and `set`) is discouraged in Python 3. The builtin types are more aggressively optimized, and will return an instance of the base type in cases where you would expect an instance of the subclass. For `dict` this is only an issue when you call `copy()`, which always returns a `dict` instead of the subclass.

It is recommended to inherit from the base classes in the `collections` module instead - in this case `collections.UserDict`

leycec replied:
> Inheriting from the built-in collection types (`dict`, `list`, and `set`) is discouraged in Python 3.

Guido, _what have you done._

Thanks for the `stackoverflow`ish high-level synopsis. I didn't know _any_ of that, suggesting I should rescind my Pythonista membership. If I could give you upvote karma, you'd be chumming it up with [Alex Martelli](http://stackoverflow.com/users/95810/alex-martelli).

> `collections.UserDict`

Where have you been my entire Python development life? It shall be so.

codewarrior0 replied:
> Is dict.copy() really called frequently enough to warrant this sort of aggressive optimization?

Probably not, but that's just one part of it. The real winners are probably list catenation, set intersection, difference, dict update and the like.

leycec replied:
> The real winners are probably list catenation, set intersection, difference, dict update and the like.

Right. We definitely want all of those to be optimized to the hilt. `</sigh>`

> ...it will call the C-level functions like `PyDict_Next` when possible, instead of going through the Python method lookup and getting `iter(obj)`...

It's a rational change from the limited perspective of optimization gains. But this is CPython, in which multithreading – the most primitive building block of parallelization, available in most modern languages out-of-the-box – is infeasible. <sup>...for all intents and purposes.</sup>

**Urgh.** Ruby definitely got it right on this one. Ruby also has a variety of working JIT interpreters coming online (e.g., B3, Rubinius). _Oh, well._

codewarrior0 commented on PyInstaller/building/utils.py:
```diff
@@ -391,32 +391,54 @@ def _rmtree(path):
+        List of `TOC`-style 2-tuples whose:
+        * First element is the absolute or relative path of a target file.
+        * Second element is the absolute or relative path of the corresponding
+          source file to be copied to this target file.
```
These docstrings look nice. We should formally adopt Numpydoc as the project's docstring format.

leycec replied:
Embrace the Numpydoc, for it is... mathy! And ASCII art.

The [Google alternative](http://bwanamarko.alwaysdata.net/napoleon/format_exception.html) isn't half-bad either, if a little terse and distinctly lacking in ASCII art. The default Sphinx format, though... just **"Ugh."** A flat list of colon-prefixed and -suffixed items in which both parameter and return types are optional (and hence often omitted, despite being the most important metadata to be documented) is displeasing to this eye and brain.

htgoebel replied:
Since we are aiming to convert the documentation to sphinx, I assume sphinx format is the only choice.

virtuald replied:
sphinx has a plugin to allow parsing NumPy and google docstrings.

htgoebel replied:
Good to know.

codewarrior0 commented on PyInstaller/building/build_main.py:
```diff
@@ -389,71 +407,73 @@ def assemble(self):
-                if node_type not in VALID_MODULE_TYPES:
+
+                #FIXME: This is... wierd. What's this about again?
+                # If this module is not a module at all, permanently ignore it.
```
I believe this is because `findNode()` is allowed to return things like `MissingModule` or `InvalidCompiledModule`, which we are not interested in running hooks for. Mostly, `MissingModule` can be created for a name that is imported from a module, as in `from foo import bar`, so we don't want those.

leycec replied:
Oh, yes. Quite right. I thought `VALID_MODULE_TYPES` meant `VALID_PYTHON_MODULE_TYPES` (e.g., `type(sys)`) rather than `VALID_MODULEGRAPH_MODULE_TYPES`.

This is what happens when people don't read documentation.

codewarrior0 commented:
> test_pandas_extension() appears to be the only failing test. 

Yep, it's not relevant here. See #1848 to find out which oddly-phrased yet valid import statement is making it choke.

Other than one or two minor nitpicks, this looks good to me. Better than I would have done it, in fact.

matysek commented:
@leycec Could you please backport from #1651 these tests?
- test_pil_no_tkinter
- test_pil_tkinter_bundled
Tests assume having defined `excludedimports` for PIL hook.

If they both pass then this pull request is in a good shape.

Thanks for you effort.

leycec commented:
> See #1848 to find out which oddly-phrased ~~yet valid~~ import statement is making it choke.

**Super intense.** I especially relished Hartmut's INTP-like [exhaustive rebuttal](https://github.com/dateutil/dateutil/pull/202#issuecomment-192352290) in the corresponding [`dateutil` issue](https://github.com/dateutil/dateutil/pull/202#issuecomment-192352290), and am delighted that he single-handedly bested this dragon.

It's unfortunate, though, that Python accepts this seemingly erroneous "multi-recursive import." Is this edge-case something ModuleGraph could be reasonably refactored to detect?

leycec commented:
> @leycec Could you please backport from #1651 these tests?

Brilliant idea, as ever. **Let's do this.**

htgoebel commented:
Re https://github.com/pyinstaller/pyinstaller/pull/1858#issuecomment-193076398
> Is this edge case something that ModuleGraph could be reasonably refactored to detect?

I'd spend *no* time adding support for *crap* code. This very code in dateutil did only work by chance. Simply using a global variable would have broken that code anyway.

leycec commented:
> I'd spend _no_ time adding support for crap code.

**Heh.** You're right, of course. And thanks for diagnosing that issue, too. You took one for the team.

leycec commented:
**Partially done!** codewarrior0's stalwart recommendations are all in, including:

* `dict` superclass **->** `UserDict`.
* `code` property **->** `scope`.

**Up next:** backporting of matysek's instructive `excludedimports` tests.

leycec commented:
Incidentally, I've begun having UPX difficulties on a local Linux machine. `py.test` now locally fails on most tests with stderr resembling:

```
LOADER: archivename is /tmp/pytest-of-leycec/pytest-5/test_tkinter_onedir_0/dist/pyi_lib_tkinter/pyi_lib_tkinter
LOADER: pyi_arch_check_cookie failed
LOADER: archivename is /tmp/pytest-of-leycec/pytest-5/test_tkinter_onedir_0/dist/pyi_lib_tkinter/pyi_lib_tkinter.pkg
LOADER: Cannot open archive: /tmp/pytest-of-leycec/pytest-5/test_tkinter_onedir_0/dist/pyi_lib_tkinter/pyi_lib_tkinter.pkg
Cannot open self /tmp/pytest-of-leycec/pytest-5/test_tkinter_onedir_0/dist/pyi_lib_tkinter/pyi_lib_tkinter or archive /tmp/pytest-of-leycec/pytest-5/test_tkinter_onedir_0/dist/pyi_lib_tkinter/pyi_lib_tkinter.pkg
```

I assume moving `upx` out of my `${PATH}` would probably fix this, but I wonder if there's a cleaner way. Perhaps I should post a belligerent issue! `</sigh>`

htgoebel commented:
What I'm missing? Why would you need an additional "namespace" level?

`importlib_load_source` returns the *module* of the hook. And each module has it's own namespace already (`mod.__dict__`). And we do not care about whatever elements are in the module's namespace - except for theses few we use (hiddenimports, datas, etc.). And when calling the `hook` functions, it is executed in the modules scope. I can not seen where there is a conflict.

BTW: Since `importlib_load_source` returns an instance of class *module*, I'd consider naming it `scope` plain wrong.

htgoebel commented on PyInstaller/building/build_main.py:
```diff
@@ -389,71 +407,72 @@ def assemble(self):
+            if not hooked_module_names:
+                break
 
+            # Else, prevent run post-graph hooks from being run again.
```
This comments meaning differs from the old one, which says:
> ... its not necessary apply then again

Anyway I wonder how ever the hooks could be run twice for one program. All this is done in `Analyse.assemble()` which is run exactly once per program.

htgoebel replied:
please ad "... in one of the next loop iterations"

leycec replied:
You got it, boss. :construction_worker:

htgoebel commented on PyInstaller/building/imphook.py:
```diff
@@ -12,18 +12,453 @@
+        ModuleHookCache._cache_id_next += 1
+
+
+    def cache_hook_dirs(self, hook_dirs):
```
private function, prefix with `_`

htgoebel replied:
Do we really need a function for this? It is called only once and it if `ModuleHookCache()` would accept the hooks-dir, it is useless.

leycec replied:
> Do we really need a function for this?

Eh, maybe not? This frost-bitten gangrenous limb was inherited from the decrepit `HooksCache` class. I'm all for simplicity. I'll chop this up.

htgoebel commented on PyInstaller/building/imphook.py:
```diff
@@ -12,18 +12,453 @@
+            self.cache_hook_dir(hook_dir)
+
+
+    def cache_hook_dir(self, hook_dir):
```
private function, prefix with `_`

htgoebel commented on PyInstaller/building/imphook.py:
```diff
@@ -12,18 +12,453 @@
+                self[module_name] = [module_hook]
+
+
+    def uncache_modules(self, module_names):
```
I find the name missleadings. Why not just call it `remove`?

htgoebel replied:
I'd prefer the function signature being `(self, *module_name)`

leycec replied:
> I find the name missleadings. Why not just call it `remove`?

...like, because, and stuff? Very well. `remove_modules` it is.

> I'd prefer the function signature being `(self, *module_name)`

A sound suggestion. Variadic modules for the win. It shall be.

htgoebel commented on PyInstaller/building/imphook.py:
```diff
@@ -12,18 +12,453 @@
+            List of all fully-qualified module names to be removed.
+        """
+
+        for module_name in set(module_names):  # Eliminate duplicate entries.
```
Eliminating duplicated here is unnecessary optimization IMHO.

leycec replied:
Yes, I suppose that's true. To the sword with it!

htgoebel commented on PyInstaller/building/imphook.py:
```diff
@@ -12,18 +12,453 @@
+        module_name,
+        hook_filename,
+        hook_module_name_prefix,
+    ):
```
Please put parameters in one or two lines.

leycec replied:
Jawohl!

htgoebel commented on PyInstaller/building/imphook.py:
```diff
@@ -12,18 +12,453 @@
+        property has yet to be accessed.
+    """
+
+    ## Private methods
```
`__init__` is not a private method. I'd suggest simply skipping such comments.

htgoebel commented on PyInstaller/building/imphook.py:
```diff
@@ -12,18 +12,453 @@
+            (e.g., `datas`) with the `format_binaries_and_datas()` helper.
+        """
+
+        # Note that the passed module graph is already a weak reference,
```
Is it? Really? Does it have to be for the ModuleHook to work correctly? If so, you'd better `assert? it.

leycec replied:
> Is it? Really? Does it have to be for the ModuleHook to work correctly?

Strictly speaking, no. But it _really_ should be. As you suggest, it shall be `asserted`.

htgoebel commented on PyInstaller/building/imphook.py:
```diff
@@ -12,18 +12,453 @@
+        self.module_name = module_name
+        self.hook_filename = hook_filename
+
+        # Name of the in-memory module fabricated to refer to this hook script.
```
Please be more verbose on why we need this. Eg.
> Each import of a hook must get it's own fresh scope, otherwise artefacts from the last run may change the results. So the hooks gets a new name on each import, so the import machinery will not reuse the existing scope.

Assuming, this is the reason, of course.

OTOH 1: Is "otherwise artefacts from the last run may change the results" correct at all? It may only be relevant when having several `Analysis()` in one .spec-file (e.g. for MERGE) and in the test-suite. But what could change between two hook runs? Of course the graph may change due to e.g. excluded imports or different pathex.

OTOH 2: Couldn't "artefacts from the last run" be even beneficial, since we could e.g. cache results from `collect_submodules` or from the even more expensive evaluations?

My sense says we'd better assume "otherwise artefacts from the last run may change the results" is correct.

leycec replied:
> Please be more verbose on why we need this.

Verbosity is my middle name. <sup>It's actually William.</sup>

We need this because, without this, **tests fail.** This was actually the most difficult aspect of this pull request to debug, and I should have documented why in detail. It took me several days to eventually realize what was happening. _Which really sucked._

So, here's what was happening. Tests would individually succeed but fail when run as a group. For every test `test_poo()`, the corresponding:

* `test_poo()[onedir]` test would succeed both when run individually _and_ when followed by a `test_poo()[onefile]` test.
* `test_poo()[onefile]` test would succeed _only_ when run individually. When preceded by a `test_poo()[onedir]` test, a `test_poo()[onefile]` test would **always fail** with a `FileNotFoundError` raised by the `PyInstaller.building.utils.format_binaries_and_datas()` helper function.

Hence, output resembling:

```
$ py.test -k 'test_poo[onedir]'
tests/functional/test_libraries.py::test_poo[onedir] PASSED
$ py.test -k 'test_poo[onefile]'
tests/functional/test_libraries.py::test_poo[onefile] PASSED
$ py.test -k test_poo
tests/functional/test_libraries.py::test_poo[onedir] PASSED
tests/functional/test_libraries.py::test_poo[onefile] FAILED
```

_wat?_ This is where I began tearing my thinning hair out.

To explain this heinousness, let's first talk about the `format_binaries_and_datas()` function. This function receives as input a list of 2-tuples `(source_dir_or_glob, target_dir)` and returns as output a list of 2-tuples `(target_file, source_file)`. A list returned by this function _cannot_ be passed back to this function, as the elements of the 2-tuples of the former are in the reverse order of the 2-tuples of the latter. Doing so reliably raises the `FileNotFoundError` error raised by the failing test above. In pseudo-Python:

```
>>> output = format_binaries_and_datas([('/usr', 'poo')])
>>> output
[('poo', '/usr')]
>>> format_binaries_and_datas(output)
FileNotFoundError: Path or glob "poo" not found or matches no files.
```

So far, so good. Since the failing test above raises the same `FileNotFoundError`, it stood to reason that the `format_binaries_and_datas()` function was receiving a list that had _already_ been returned by a prior call to that function. But how could that be? That function is only called twice in the codebase: once by **(A)** `Analysis.__init__()` to handle additional `binaries` and `datas` defined by user `.spec` files and once by **(B)** `ModuleHook._sanitize_hook_attrs()` to handle `binaries` and `datas` defined by the current hook script.

So, what gives? The `format_binaries_and_datas()` function is _never_ called twice for the same hook script. That function should _never_ receive a list output by a prior call to that function. **And then I realized with horror.**

Previously, each hook script was _always_ loaded into the same in-memory module `PyInstaller_hook_{modname}` by the `importlib_load_source()` helper function. If this function were (for whatever reason) called twice with the same module name in the same Python process, the second call would attempt to reload the previously loaded hook script back into that module – which already exists. Is that problematic?

**It is.** At least under Python 3, where `importlib_load_source()` reduces to `importlib.machinery.SourceFileLoader(modname, pathname).load_module()`. For whatever reason, that logic behaves erratically when passed the name of a module that's already been loaded. I didn't delve into the specifics, but this logic appears to silently ignore attempts to reload already loaded modules. _Isn't that special._

If `py.test` runs the `test_poo()[onedir]` and `test_poo()[onefile]` tests from the same Python process (as it appears to) rather than forking a new process for each (as I assumed it did), these two tests will share the same set of in-memory modules. Due to the above `importlib_load_source()` behaviour, each `ModuleHook` object loaded for the second `test_poo()[onefile]` test will simply reuse the previously cached hook script module rather than reloading that module.

Is there a point to any of this? **Yes.** _Sort of._ Any `datas` and `binaries` attributes previously set by the first `test_poo()[onedir]` test would then be reused by the second `test_poo()[onefile]` test – except those attributes were already sanitized by `format_binaries_and_datas()`. The second `test_poo()[onefile]` test doesn't know that, of course, so it tries to resanitize them.

And this is why I didn't document this in the code in detail.

htgoebel replied:
Puh! What a debug orgy. This is really driving one nuts. Thanks a lot for this long explanation. 

But I see a chance of getting rid of it. You say, this is caused by data already sanitized by format_binaries_and_datas() being in the hook. Now after eliminating the `scope` namespace and implementing `__getattr__` (see [here](https://github.com/pyinstaller/pyinstaller/pull/1858#issuecomment-194051310)), the sanitized data is not pushed into the hook's namespace and thus this prblem should not occur anymore. Could be worth trying.

htgoebel commented on PyInstaller/building/imphook.py:
```diff
@@ -12,18 +12,453 @@
+    ## Properties
+
+    @property
+    def scope(self):
```
As already written in the main comments, this name is wrong, since is returns a module. And IMHO this level is useless anyway.

htgoebel commented on PyInstaller/building/imphook.py:
```diff
@@ -12,18 +12,453 @@
+
+        # Convert list attributes in which order is insignificant to sets,
+        # defaulting undefined lists to empty sets.
+        for attr_name in ('excludedimports',):
```
IMHO this loop does not simplify the code.
I admit, that this should be the same scheme as above. But then
```
for attr_name, func in (
   ('binaries', format_binaries_and_datas),
   ('datas', format_binaries_and_datas),
   ('excludedimports', lamba a: a),
   ('hiddenimports', lamba a: a),
   ):
  ...
```
would be terse and more concise IMHO.

leycec replied:
A sound suggestion, as ever. I debated this back and forth with myself, but didn't think of aggregating everything into one giant `for` loop with lambdas.

It is clever. It shall be!

htgoebel commented on PyInstaller/building/imphook.py:
```diff
@@ -12,18 +12,453 @@
+        # Default all remaining undefined lists to empty. Since order is
+        # significant in these lists, these lists must not be converted to sets.
+        for attr_name in ('hiddenimports',):
+            if  getattr(self.scope, attr_name, None) is None:
```
Why is this different to the code above?

leycec replied:
Because laziness and ineptitude. They are [virtues](http://blog.teamtreehouse.com/the-programmers-virtues), I tell you!

<sup>...I'll fix this.</sup>

htgoebel commented on PyInstaller/building/imphook.py:
```diff
@@ -12,18 +12,453 @@
+
+    def _process_hook_func(self):
+        """
+        Call this hook's `hook()` function if defined _or_ noop otherwise.
```
"_or_ noop otherwise" is a matter of course IMHO.

leycec replied:
No, you're quite right. I reused that noxious `_or_ noop otherwise.` phrase in a few other places too. I'll ceremonially gut them all.

htgoebel commented on PyInstaller/building/imphook.py:
```diff
@@ -12,18 +12,453 @@
+        self.scope.hook(hook_api)
+
+        # Process all attributes of this API modified by the prior call.
+        self.scope.datas.update(set(hook_api._added_datas))
```
This changes the values *within* the module's scope.

leycec replied:
Yup! Isn't life grand?

The `_sanitize_hook_attrs()` method has already converted the `datas` attribute for this hook from a list of hook-style 2-tuples into a set of `TOC`-style 2-tuples. The above statement then updates that set with the additional `datas` added by this hook's `hook()` function (if any).

htgoebel replied:
And this is causing [your pain](https://github.com/pyinstaller/pyinstaller/pull/1858/files#r55469955) :-\

htgoebel commented:
Re the namespaces, (also see this [comment](https://github.com/pyinstaller/pyinstaller/pull/1858/files#r55349317)): At https://github.com/pyinstaller/pyinstaller/pull/1858/files#diff-ad80caf8573f1614b16a122d9c756b7cR288 we currently *change* the variables inside the hooks scope. Esp. we change the structure of `binaries` and `datas`, *before* `hook()` is called. This may break hooks which update these variables.

I'd rather *pull* relevant variables out of the hook. This is why I suggested using `__getattr__`.

htgoebel commented on PyInstaller/configure.py:
```diff
@@ -83,6 +83,8 @@ def _get_pyinst_cache_dir():
     return cache_dir
 
 
+#FIXME: Rename to get_hooks_official_dir().
```
Should be named `get_official_hooks_dir`.

leycec replied:
I shake my fist at human-readable function names! Nonetheless, I'll do it.

htgoebel commented on PyInstaller/hooks/pre_find_module_path/hook-site.py:
```diff
@@ -23,6 +23,9 @@
 from PyInstaller import PACKAGEPATH
 
 def pre_find_module_path(api):
+    #FIXME: For reusability, shift this path into a new
```
s/path/function/

htgoebel commented on PyInstaller/building/build_main.py:
```diff
@@ -374,10 +374,28 @@ def assemble(self):
+        # listed before and hence take precedence over custom hooks.
+        module_hook_dirs = [get_importhooks_dir()] + self.hookspath
+
+        # Lazily defer loading hook scripts until their first access.
```
This does not "Lazily defer", but prepooulate the cache - which itself is in charge of lazy defer.

leycec replied:
You say tomato, I say rutabaga. But, yes. "Prepopulate" is probably a more accurate verb. Yay!

htgoebel commented on PyInstaller/building/build_main.py:
```diff
@@ -389,71 +407,72 @@ def assemble(self):
-        # 4. The infinite 'while' loop ends when:
-        #    a. hooks cache is empty
-        #    b. no new hook was applied in the 'while' iteration.
+        # Post-graph hooks are now run until no such hooks remain. Specifically,
```
What does this "now" mean?

leycec replied:
"Below," in this case. O.K., let's see if I can clarify this up.

htgoebel commented on PyInstaller/building/build_main.py:
```diff
@@ -389,71 +407,72 @@ def assemble(self):
+                break
 
+            # Else, prevent run post-graph hooks from being run again.
+            module_hook_cache.uncache_modules(hooked_module_names)
```
I'd move this in front of the `if not hooked_module_names: break` line. This makes it easier to understand. (Yes, I know, if `not hooked_module_names`, calling this function is a no-op.)

Or write it the other way round: `if hooked_module_names: ... uncache() else: break`.

leycec replied:
> I'd move this in front of the `if not hooked_module_names: break` line. This makes it easier to understand.

Quite right. It saddens me that Python has no `do ... while` loop construct, which would simplify things even further.

htgoebel replied:
> that Python has no do ... while loop construct

Yes, this is one of the rare missing in Python.

htgoebel commented on PyInstaller/building/imphook.py:
```diff
@@ -12,18 +12,453 @@
+    This identifier is incremented on each instantiation of a new
+    `ModuleHookCache` to isolate in-memory modules of lazily loaded hook script
+    code objects in that cache to the same cache-specific namespace, preventing
+    edge-case collisions with existing in-memory modules in other caches.
```
1) I do not understand this explanation, see https://github.com/pyinstaller/pyinstaller/pull/1858/files#r55348155 for what I understood about this.

2) General: This kind of documentation should be but into comments, not into strings. The strings will go into the byte-code but are useless and thus only wasting disk-space and memory. Just because it is a tripple-quoted string does not make it a [doc-string](http://docs.python-guide.org/en/latest/writing/documentation/#code-documentation-advice).

leycec replied:
> The strings will go into the byte-code but are useless and thus only wasting disk-space and memory.

**Ah-ha!** I know something you do not. I now feel smug. :wink: 

In all seriousness, the Python bytecode compiler [optimizes away](https://www.python.org/dev/peps/pep-0224) this string. It doesn't actually generate any additional code. It's a noop, sadly.

> Just because it is a tripple-quoted string does not make it a doc-string.

While it's true that Python itself ignores this string, [Sphinx](https://stackoverflow.com/a/22768779/2809027) and most intelligent IDEs (e.g., PyCharm) parse this string in a semantically meaningful way.

But – no big deal! I'd be happy to convert this into a brute-force multiline comment, if you're keen. That's what Vim `gcc` is for. <sup>Praise be to [Bram Moolenar](http://www.moolenaar.net), for he is holy.</sup>

htgoebel replied:
> the Python bytecode compiler optimizes away this string

I'd bet my grandmother on it: older versions did not. Now *you* beat me ;-) Thanks for enlightening me. So the string-style is the better one: easier to read and easier to maintain.

htgoebel commented on PyInstaller/building/imphook.py:
```diff
@@ -12,18 +12,453 @@
+        # Curiously, calling "super(UserDict, self).__init__()" here instead
+        # subsequently results in exceptions resembling:
+        #     AttributeError: 'ModuleHookCache' object has no attribute 'data'
+        UserDict.__init__(self)
```
Why not using super here?

leycec replied:
Your first intuition was right, actually. This _should_ be using `super`.

I initially thought the `UserDict` class was simply b0rked. Only after pushing that comment did I realize (with the sort of growing horror that gnaws at your gut) the truth: I'd originally refactored this code from the older `HooksCache` class, **which is subtly broken:**

```
class HooksCache(dict):
    def __init__(self, hooks_dir):
        # This doesn't actually do anything. This should instead be:
        #super(HooksCache, self).__init__()
        super(dict, self).__init__()
```

I now know: never trust the lies of old code.

htgoebel commented on PyInstaller/building/imphook.py:
```diff
@@ -12,18 +12,453 @@
+
+        # String unique to this cache prefixing the names of all in-memory
+        # modules lazily loaded from cached hook scripts.
+        self._hook_module_name_prefix = 'PyInstaller_hooks_{}_'.format(
```
I'd additionally prefix the prefix with `__` to mark it private.

leycec replied:
Two underscores! Now you're playing with power. It shall be so, for it is a healthy idea.

htgoebel commented on PyInstaller/building/build_main.py:
```diff
@@ -374,10 +374,28 @@ def assemble(self):
+
+        # Lazily defer loading hook scripts until their first access.
+        module_hook_cache = ModuleHookCache(module_graph=self.graph)
+        module_hook_cache.cache_hook_dirs(module_hook_dirs)
```
Why not passing `module_hook_dirs` to `ModuleHookCache()` already?

leycec replied:
Because: **hand-waving.**

Actually, I'll add an optional parameter to `ModuleHookCache()` supporting that. Have a Swedish lollipop. :lollipop:

htgoebel commented on PyInstaller/building/imphook.py:
```diff
@@ -12,18 +12,453 @@
+
+            # If at least one hook object has been cached for this module,
+            # append this hook object to the existing list of such objects.
+            if module_name in self:
```
Can't use `setdefault()` here`

leycec replied:
`dict.setdefault()` – _that_ was it! I knew this conditional could be optimized away somehow, but couldn't remember how to go about.

My first thought was `collections.defaultdict`. Which was an awful thought, so I ignored it. Thanks for the patient reminder.

htgoebel commented on PyInstaller/building/imphookapi.py:
```diff
@@ -85,11 +85,15 @@ def module_graph(self):
         return self._module_graph
 
+    #FIXME: Remove this property and publicize the "_module_name" attribute to
+    #"module_name". This attribute is intended to be modifiable by hooks.
```
> This attribute is intended to be modifiable by hooks.

Really? Where is this documented?

leycec replied:
Nowhere, Hartmut. Absolutely nowhere. Like all good things in life, you were just expected to know. (_Somehow._)

Nonetheless, it's the case. See this logic in the `PyiModuleGraph._safe_import_module()` method:

```
                # Respect method call changes requested by this hook.
                module_basename = hook_api.module_basename
                module_name = hook_api.module_name
```

htgoebel replied:
You caught me.

leycec commented:
_Oh, uncaring Gods._ So much commentary. This is good, but will delay things a bit. Such is peer review!

Thanks for the in-depth code review, Hartmut. It's back to the sad code grindstone. :dizzy_face: 

> And we do not care about whatever elements are in the module's namespace - except for theses few we use (hiddenimports, datas, etc.). And when calling the hook functions, it is executed in the modules scope.

**I get it, now.** You propose conditionally exposing _only_ the subset of hook attributes and functions required externally via `ModuleHook.__getattr__()` – say, `datas`, `binaries`, `hiddenimports`, `excludedimports`, `hook()`, `pre_find_module_path()`, and `pre_safe_import_module()`. Did I miss any? (Let me guess: _Yes_.)

This would replace the existing `scope` property entirely, preventing external callers from accessing hook script internals unsafely. Encapsulation, of a sorts.

O.K., I can [get behind that in the morning and plow](https://www.youtube.com/watch?v=l7yuTR8r6QM). While I usually prefer the direct approach to `__getattr__()` shenanigans, a case can probably be made for black magic here. For completeness, I'd probably also want to implement `__setattr__()` for the hook attributes. <sup>Someone will want that. I know someone.</sup>

How do the two musketeers codewarrior0 and matysek feel about this? I'd like a bit of confirmation before I run half-cocked into the jungle again.

codewarrior0 commented:
> How do the two musketeers codewarrior0 and matysek feel about this?

I have to reverse myself now and agree with @htgoebel. The extra layer of indirection isn't needed if we only pull out the names we care about from the hook module.

> @codewarrior0 Help! Help! I summon thee.

Hum. `_FORTIFY_SOURCE` was added after a recent discussion. Pardon me while I go over that again - maybe it's defined by default in some systems' header files?

leycec commented:
> I have to reverse myself now and agree with @htgoebel. The extra layer of indirection isn't needed if we only pull out the names we care about from the hook module.

**Concensus has spaketh.** This will take a bit of jiggering to jiggle into place and document properly, but I can certainly see the sense it in. Onward, code brethren!

> Hum. `_FORTIFY_SOURCE` was added after a recent discussion. Pardon me while I go over that again - maybe it's defined by default in some systems' header files?

_Yesss._ Gentoo and `gcc` are vestigially tied at the hip. I wouldn't be surprised if the default Gentoo profile already silently forces `_FORTIFY_SOURCE`. (Somehow.)

**Indeed,** this appears to be the case. From this 2009 (!) [Gentoo security advisory](https://archives.gentoo.org/gentoo-dev/message/8efa8bf5aad1fd2d724f4ba001b26ff7):

> not to be out done, `gcc-4.3.2-r3` will include changes like some other distros 
> are now carrying:
>  - the `-Wformat-security` flag is enabled by default
>  - the `-D_FORTIFY_SOURCE=2` flag is enabled by default
>
> if you dont want this stuff, you can use the flag `-Wno-format-security` and the 
> flag `-U_FORTIFY_SOURCE` respectively
> -mike

It's unfortunate that `gcc` can't tolerate multiple `-D_FORTIFY_SOURCE=2` flags being passed. What the heck is that command-line fragility?

The "...like some other distros are now carrying" phrase above suggests this isn't just a Gentoo complication. Similarly hardcore distros (e.g., Arch, Slackware) might be running similar defaults. This is my head hitting the desk.

htgoebel commented:
> Oh, uncaring Gods. So much commentary

I'm sorry about this. But it is a big change at a central place. So I'm reviewing it in detail.

Anyway: I'm glad you take the courage of implementing all this stuff! We need need guy like you, carving out old code and bringing in new concepts. Otherwise PyInstaller would be legacy soon.

htgoebel commented:
Hi @leycec did I discourage you?

htgoebel commented:
@leycec pinging.

leycec commented:
> Hi @leycec did I discourage you?

_"Never, fiend!"_ Err, sorry. Too much late-night Baldur's Gate. I mean:

_"Never, friend!"_ Your incisive critiques have only emboldened my desire to – against all odds – end this once and for all. This is complicated, of course, by my newfound inability to [locally recompile](https://github.com/pyinstaller/pyinstaller/pull/1858#issuecomment-195161021) the bootloader. Assuming that hasn't been fixed yet, fixing that will probably be my next order of business. Lazy Hook Part II will have to wait, I'm afraid.

On the bright side, I've nearly finished my revisions for this pull request. Real life intruded. (_**Read:** wild squirrels inside our roof. I shall eat them all._)

codewarrior0 commented:
> fixing that will (probably) be my next order of business

Let me see what I can come up with...

```
diff --git a/bootloader/wscript b/bootloader/wscript
index 7a127c2..3e3cd07 100644
--- a/bootloader/wscript
+++ b/bootloader/wscript
@@ -393,6 +393,8 @@ def configure(ctx):
         # in various string and memory manipulation functions.
         # TODO What other platforms support _FORTIFY_SOURCE macro? OS X?
         if is_linux and ctx.env.CC_NAME == 'gcc':
+            # Undefine existing _FORTIFY_SOURCE in case the compiler defines it already
+            ctx.env.append_value('CFLAGS', '-U_FORTIFY_SOURCE')
             ctx.env.append_value('DEFINES', '_FORTIFY_SOURCE=2')

     if is_win:
```

I'm not a Gentoo user so I don't know for sure that it works... but it should.

leycec commented:
@codewarrior0? More like code**savior**0! Let's give that a rimshot.

leycec commented:
@codewarrior0 Interestingly, OS X's `clang` [appears to have supported](http://permalink.gmane.org/gmane.comp.compilers.clang.devel/2442) the `_FORTIFY_SOURCE` macro since 2008 (and probably earlier). It's been eight long years. This support _should_ actually work now. :octopus:

codewarrior0 commented:
Ah, I think you missed a line. The `-U_FORTIFY_SOURCE` option in CFLAGS only undefines the macro if it is already defined, to avoid the "macro redefinition" error. The `_FORTIFY_SOURCE=2` in DEFINES is still needed to actually enable the macro!

leycec commented:
**Woops.** That is terrible. By the gurgling of my empty stomach, I see yet again that coding and fasting don't mix. Thanks for the erudite clarification!

leycec commented:
**Done,** baring the inevitable test failures.

@matysek As requested, your excellent PIL tests have been [backported](https://github.com/pyinstaller/pyinstaller/pull/1858/commits/16fbdb657d89ff14d754442539b1be09dffc93d4) from #1584. They succeed locally, but let's see how Travis feels about that.

@htgoebel As requested, your excellent recommendations have been noted and duly ignored. _Ha!_ O.K., I actually implemented everything – including refactoring away the `scope` (formerly `code`) property with black `__getattr__()` magic. For parity, I've also defined similar `__setattr__()` and `__delattr__()` magic. Hook attributes are now safely gettable, settable, and deletable.

leycec commented:
_"Curiouser and curiouser!"_ **[*](https://www.cs.cmu.edu/~rgs/alice-II.html)**

All tests succeed _except_ matysek's backported `test_pil_no_tkinter()` test, which fails with `ERROR: Module _tkinter is bundled.`. Since this test locally succeeds _and_ since I preserved matysek's `_process_excluded_imports()` method without modification, I... I don't even anymore.

![Serenity now!](http://memecrunch.com/meme/59H5/serenity-now-squirrel/image.png?w=400&c=1)

The error message means that the `_tkinter` module was _not_ excluded. It should have been. Interestingly, the `hook-PIL` hook doesn't explicitly exclude `_tkinter`; it assumes that only Tkinter itself (which `hook-PIL` does explicitly exclude) imports `_tkinter`. Together, this implies that either:

1. The current `excludedimports` approach is broken, which probably has nothing to do with lazy hook loading and this pull request.
1. The current `excludedimports` approach works, but something other than Tkinter is importing `_tkinter` under Travis – which is feasible, if a bit weirdo. That also has nothing to do with lazy hook loading and this pull request.

Does this test also fail under the older non-lazy-loading `HooksCache` and `ImportHook` classes? I suspiciously suspect it does.

I'd prefer to stomp this bug into a bloody smear, but... doing so is probably non-trivial. While important, import exclusion is outside the mandate of this pull request. Perhaps we could defer that for another volunteer day?

leycec commented:
_"Curiouser and even curiouser!"_

I couldn't help myself, so I debugged a bit. I'm now convinced that **1.** above is the case: **the current `excludedimports` approach is broken.** Specifically, I switched the order of imports in the backported `test_pil_no_tkinter()` – and it now fails locally! What kind of nondeterministic wackiness is _this_?

In #1651, I give an [alternate `excludedimports` solution](https://github.com/pyinstaller/pyinstaller/pull/1651#issuecomment-192550183) requiring only ten lines of code. Coincidentally, that code requires lazy hook loading. We have lazy hook loading. Since this now seems like low-hanging fruit, I'm going to try ripping out the existing `ModuleHook._process_excluded_imports()` approach and replace it with a `PyiModuleGraph._safe_import_hook()` approach.

This has _nothing_ to do with lazy hook loading or this pull request. I should really save this for a subsequent pull request, but... _what could it hurt?_ <sup>`crickets`</sup>

htgoebel commented on PyInstaller/building/imphook.py:
```diff
@@ -12,18 +12,573 @@
+            ('excludedimports', set),
+
+            # Preserve lists in which order is significant as lists.
+            ('hiddenimports', list),   # order of importation is significant
```
Huh, order is significant? Why?

leycec replied:
**Hooks, mostly.**

The mere act of loading a hook script or calling a hook function defined by a hook script can have side effects conditionally dependent on module importation order. Since module importation triggers both the loading of hook scripts _and_ calling of hook functions defined by hook scripts, it follows that module importation itself can have these sorts of side effects.

_"Say wat-wat?_", you are thinking. I once thought like you. Now, consider two worst-case hook scripts `hook-Yog.Sothoth.py` and `hook-Shub.Niggurath.py` defined as follows:

* In `hook-Yog.Sothoth.py`:

        # I am evil hook script! I do what should not be done.
        from PyInstaller.depend import analysis
        if getattr(analysis, '_IS_GREAT_OLD_ONE_IMPORTED', False):
            hiddenimports = ['Azathoth', 'Ghatanothoa']
        else:
            analysis._IS_GREAT_OLD_ONE_IMPORTED = True

* In `hook-Shub.Niggurath.py`:

        # I am evil hook script also! I too do what should not be done.
        from PyInstaller.depend import analysis
        if getattr(analysis, '_IS_GREAT_OLD_ONE_IMPORTED', False):
            excludedimports = ['Nyarlathotep', 'Yig']
        else:
            analysis._IS_GREAT_OLD_ONE_IMPORTED = True

That's an extreme example. A saner example would probably have these two hook scripts define post-graph `hook()` functions conditionally adding hidden and/or excluded imports conditionally depending on whether or not certain modules had already been imported (e.g., by querying the current module graph).

In any case, order of importation is significant. Hooks are interpreted by the current Python process and hence have the keys to the PyInstaller kingdom. Although pathological hooks cannot be prevented from behaving pathologically, PyInstaller can at least ensure that it does its best to "do the right thing."

Of course, no well-behaved hook script should have these sorts of side effects. In the ideal world, _all_ hooks would refrain from modifying anything outside their jurisdiction (i.e., the local script in which they're defined) and would _not_ conditionally depend on module importation order. In the actual horrible world, hooks can anarchistically do anything they like – including monkey patch low-level PyInstaller and even lower-level Python operation.

Which means that somewhere, sometime, someone will do it. And they'll be really smug about it. And, when it burns down their multi-million dollar NSA datacenter, they'll post an accusatory PyInstaller issue. _And it's somehow all our fault._

htgoebel commented on PyInstaller/building/imphook.py:
```diff
@@ -12,18 +12,573 @@
+        # If this is a magic attribute, set this attribute to the passed value.
+        if attr_name in _MAGIC_MODULE_HOOK_ATTR_NAMES:
+            # Validate this attribute's new value.
+            if attr_name in {'datas', 'binaries', 'excludedimports'}:
```
IMHO this is over-engineered. But if you like this ...

But if you re doing this why not add more informations to `MAGIC_MODULE_HOOK_ATTR_NAMES` so everything is defined at a single place. As of now you have three definitions: line 166, line 268 and here.

leycec replied:
> IMHO this is over-engineered.

That's valid criticism. You're probably right. Synchronizing this validation with magic attribute names and types declared elsewhere is a recipe for death soup. And I'm tired of death soup.

So:

1. I'll remove this validation.
1. I'll add magic attribute types to `MAGIC_MODULE_HOOK_ATTR_NAMES` – probably by refactoring that `set` into a `dict`.

htgoebel commented on PyInstaller/building/imphook.py:
```diff
@@ -12,18 +12,573 @@
+        # loading this hook script and then remove this attribute.
+        if attr_name in _MAGIC_MODULE_HOOK_ATTR_NAMES:
+            self._load_hook_module()
+            del self.__hook_attrs[attr_name]
```
Hm, this is asymmetric: Line 268 defines default values, but here one can delete the attribute.

You could save this `__delattr__` anyway if storing the hook attrs in `self` (e.g. `setattr('data', value)`. This is [the trick about the `__getattr__` stuff](https://docs.python.org/3/reference/datamodel.html#object.__getattr__).

leycec replied:
> Hm, this is asymmetric: Line 268 defines default values, but here one can delete the attribute.

But... isn't that the case with most instance variables initialized by an `__init__()` method?

> You could save this `__delattr__` anyway if storing the hook attrs in `self`...

I think I see what you're getting at. _Maybe?_

For each hook attribute (e.g., `self._hook_module.datas`), you propose dynamically defining an instance variable of the same name in the current `ModuleHook` instance (e.g., `self.datas`). These variables would then be publicly accessible in the standard way without having to go through the `__getattr__()` magic lookup process. Is that right? If so, I begrudgingly admit: **that works, too.**

It's also simpler and more efficient. So, I'm sold. We'll go with that. I kind of liked firewalling magic attributes from external access into a private `dict`, but we can always privatize this later if we really need to. (_We probably won't._)

htgoebel commented:
@leycec Thanks for the update. I don't have time to review it ATM. I hope I'll find some time later this weekend.

leycec commented:
>  I don't have time to review it ATM.

_Wait._ Wasn't that your review above? :weary:

leycec commented:
To no one's surprise, properly implementing `excludedimports` proved... _non-trivial._ The `ModuleGraph` class requires a bit of refactoring to support my ideal solution. Only a bit, but things escalated fast. I'd like to defer that work to my next pull request.

This means that matysek's backported `test_pil_no_tkinter()` test will continue to fail, as it currently does. I've marked this test `@xfail` for the moment. At least the failure is explicitly visible now.

So much for Lazy Hook Parts II and III. We all had dreams once. I need a bottle of hard sake. :sake:

htgoebel commented:
>> I don't have time to review it ATM.

> Wait. Wasn't that your review above?

You got me.

Sorry for the permanent nitpicking. Sometime my perfectionism is getting the better of me. I think, we should merge this as soon as you updated the pull-request to xfail this one test. Any fine-tuning can be done later. Let's finish this important step to be able to make the next.

(I'll be away in a few hours, so maybe some other @pyinstaller/core-developers can merge when leycec updated this pull-request.)

matysek commented:
I'm probably late to the party. however, could anyone offer please a summary for this long crazy story?

Thank you very much.

leycec commented:
> I'm probably late to the party. however, could anyone offer please a summary for this long crazy story?

For you, the party's always open.

Most of the discussion revolved around htgoebel's exhaustive (..._and I do mean **exhaustive**_) peer review of my proposed changes. htgoebel is a great man, but his unyielding search for quality pains this gentle soul. In theory, the radioactive fallout has finally settled. My high-level summary:

1. I originally exposed hook attributes and functions via a `ModuleHook.code` property.
1. codewarrior0 recommended I rename that, as the value exposed by that property wasn't actually a code object; it was a module object. He was right.
1. I renamed that property to `ModuleHook.scope`, one of codewarrior0's helpful alternatives.
1. htgoebel declared that to be a terrible name <sup>_(Fair enough.)_</sup> and that, more importantly, no property was needed at all. Instead, he suggested dynamically exposing the four hook attributes of interest (i.e., `datas`, `binaries`, `excludedimports`, `hiddenimports`) as instance variables of the `ModuleHook` class by overriding the `ModuleHook.__getattr__()` special method. This is clever. codewarrior0 agreed. I dutifully obeyed.
1. I implemented `ModuleHook.__getattr__()`, `ModuleHook.__setattr__()`, and `ModuleHook.__delattr__()` by isolating these hook attributes into a private `ModuleHook._hook_attrs` dictionary.
1. htgoebel declared that to be a terrible approach. <sup>_(Fair enough.)_</sup> Instead, he suggested directly exposing these hook attributes as instance variables of the current `ModuleHook` instance. This is simpler, and has the fringe benefit of allowing these hook attributes to be safely deleted without needing to override `ModuleHook.__delattr__()`. This is also clever. I agreed. I dutifully obeyed.

I also backported the two tests you requested. Since the current `excludedimports` implementation appears to be non-deterministically broken, the `test_pil_no_tkinter()` fails and has been temporarily marked as `@xfail`. I _very_ briefly attempted to replace the current `excludedimports` implementation with an alternative overriding the `ModuleGraph._safe_import_hook()` method, but quickly became bogged down in low-level `ModuleGraph` details. _The inner turmoil was real that evening._ :finnadie: 

I've stashed those changes away for a subsequent pull request. And here we stand, idling at the end of days – or at least of this pull request. **Hopefully.**

leycec commented:
> Sorry for the permanent nitpicking. Sometime my perfectionism is getting the better of me.

You are dedicated to cleverness, and the code's better for it. Still, I'd hoped to get a bit more done on the codebase before my time was up.

My time's up. The real world's calling. I regret leaving Lazy Hook Part II and III undone, but... maybe someone else can have a go at that? If not, I'll happily pick up where I left off when free time returns.

> I think, we should merge this as soon as you updated the pull-request to xfail this one test.

It is [done](https://github.com/pyinstaller/pyinstaller/pull/1858/commits/607341439a84fea54fedf40cf4e360d75577b73f).

> Any fine-tuning can be done later. Let's finish this important step to be able to make the next.

Yes! To the [future world](https://www.youtube.com/watch?v=IO7tBFTxVEw), with a heart that's full of joy.

leycec commented:
**OMFG.** `UserDict` is a bloody _old-style_ class under Python 2.7. Guido, I blame you.

leycec commented:
**All tests pass.**

Travis appears to fail but actually passes. The issue is non-determinism in the `test_pil_no_tkinter()` test. For understandable reasons, `py.test` expects tests to behave deterministically: either the test always succeeds _or_ the test always fails. Since `test_pil_no_tkinter()` has been marked as `@xfail`, `py.test` expects this test to always fail.

It usually does. Under _only_ Python 3.4 on Linux, however, this test miraculously succeeds – which `py.test` interprets as an `xpass` (i.e., unexpected success), equivalent to failure. Since this is dumb, I've temporarily marked `test_pil_no_tkinter()` as an unconditional `@skipif(True)` rather than a conditional `@xfail`.

Off and away we go.

leycec commented:
**All tests _really_ pass.**

We have ignition. :rocket:

htgoebel commented:
>> Sorry for the permanent nitpicking. Sometime my perfectionism is getting the better of me.

> You are dedicated to cleverness, and the code's better for it. Still, I'd hoped to get a bit more done on PyInstaller before my time was up.

Sorry. I learned my lesson: Next time I'll merge quicker and the refine he code.

> My time's up. The real world's calling. I regret leaving Lazy Hook Part II and III undone, but... maybe someone wiser can have a go at that? If not, I'll happily pick up where I left off when free time returns. ...as if

If you already have some code for this, I'd appreciate completing it - while giving the credits to you. What are the topics of part II and III?

leycec commented:
**The Eagle has landed.** Thanks so much for the prompt merge, Hartmut. That means alot to me... and to my vanishing free time.

> What are the topics of part II and III?

Right. Given `excludedimports`, a Part IV is also probably in order. In order of increasing difficulty:

1. Lazy Hook Part II: **unify pre-find module path hooks with post-graph hooks.** Specifically:
    1. Merge the contents of all hook scripts in the `hooks/pre_find_module_path/` subdirectory with the contents of existing hook scripts in the `hooks/` directory.
    1. Remove the  `hooks/pre_find_module_path/` subdirectory.
    1. Refactor `PyiModuleGraph` to call pre-find module path hooks by lazily loading hook scripts from the new `ModuleHooksCache` singleton.
1. Lazy Hook Part III: **unify pre-safe import module hooks with post-graph hooks.** Same procedure as in 1., replacing `hooks/pre_find_module_path/` with `hooks/pre_safe_import_module/`.
1. Lazy Hook Part IV: **refactor `excludedimports` processing to be performed _during_ rather than _after_ module graph creation.** Specifically, define a new `PyiModuleGraph._safe_import_hook()` method overriding the superclass `ModuleGraph._safe_import_hook()` method with support for `excludedimports` processing.

Parts II and III are tedious busywork, but guaranteed to succeed. Part IV is the opposite; it's non-tedious refactoring, but _not_ guaranteed to succeed. I believe this approach will successfully solve our outstanding `excludedimports` woes, but there's no way of knowing for sure until I give it a go.

Food for future thought.

matysek commented:
@leycec and @htgoebel thank you very much for this great work. I really appreciate it.

Does it mean my original attempt #1651 could be closed?

matysek commented:
I created followup issues for the Lazy Hook Parts.
